staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/values.go

```diff
@@ -0,0 +1,611 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cel
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+	"github.com/google/cel-go/common/types/traits"
+	structuralschema "k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
+	"k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model"
+	"k8s.io/kube-openapi/pkg/validation/strfmt"
+)
+
+// UnstructuredToVal converts a Kubernetes unstructured data element to a CEL Val.
+func UnstructuredToVal(unstructured interface{}, schema *structuralschema.Structural) ref.Val {
+	if schema.XIntOrString {
+		m, ok := unstructured.(map[string]interface{})
+		if !ok {
+			types.NewErr("invalid data, expected map[string]interface{} to match the provided XIntOrString schema")
+		}
+		if i, ok := m["intVal"]; ok {
+			return types.DefaultTypeAdapter.NativeToValue(i)
+		}
+		if s, ok := m["strVal"]; ok {
+			return types.DefaultTypeAdapter.NativeToValue(s)
+		}
+		return types.NewErr("invalid data, expected XIntOrString value to have either a intVal or strVal field")
+	}
+	if schema.Type == "object" {
+		m, ok := unstructured.(map[string]interface{})
+		if !ok {
+			types.NewErr("invalid data, expected map[string]interface{} to match the provided schema with type=object")
+		}
+		if schema.XEmbeddedResource {
+			// only allow access to name and generateNames of metadata by pruning all other metadata fields.
+			return &unstructuredMap{
+				value:      m,
+				schema:     embeddedSchema,
```

Comment:
can't a property with `x-kubernetes-embedded-resource: true` define its own properties in addition to standard metadata/kind/apiVersion properties? if `embeddedSchema` only includes metadata/kind/apiVersion, are those other properties usable? should we be augmenting `schema` with metadata.name / metadata.generateName / kind / apiVersion instead of replacing it entirely?

I'm not quite sure how the schema set here interacts with propSchema